Description: Check for proper index set in calls to GetDefaultPlayDevice and
 GetDefaultRecordDevice, to avoid crashes later. 
Bug-Ubuntu: https://launchpad.net/bugs/475277
Forwarded: not-needed
Origin: http://audacity.cvs.sourceforge.net/viewvc/audacity/audacity-src/src/prefs/DevicePrefs.cpp?r1=1.6&r2=1.7
diff -Nur -x '*.orig' -x '*~' audacity/src/prefs/DevicePrefs.cpp audacity.new/src/prefs/DevicePrefs.cpp
--- audacity/src/prefs/DevicePrefs.cpp	2009-11-12 21:31:24.097328400 +0100
+++ audacity.new/src/prefs/DevicePrefs.cpp	2009-11-12 21:32:06.944830051 +0100
@@ -204,10 +204,12 @@
    if (mPlay->GetCount() == 0) {
       playnames.Add(_("No devices found"));
       mPlay->Append(playnames[0], (void *) NULL);
+      mPlay->SetSelection(0);
    }
    if (mRecord->GetCount() == 0) {
       recordnames.Add(_("No devices found"));
       mRecord->Append(recordnames[0], (void *) NULL);
+      mRecord->SetSelection(0);
    }
 
    /* what if we have no device selected? we should choose the default on 
@@ -296,8 +298,11 @@
 
 wxString DevicePrefs::GetDefaultPlayDevice(int index)
 {
-   const struct PaHostApiInfo *apiinfo = Pa_GetHostApiInfo(index);
-   // get info on API
+   if (index < 0 || index >= Pa_GetHostApiCount()) {
+      return wxEmptyString;
+   }
+
+   const struct PaHostApiInfo *apiinfo = Pa_GetHostApiInfo(index);   // get info on API
    wxLogDebug(wxT("GetDefaultPlayDevice(): HostAPI index %d, name %s"), index, wxString(apiinfo->name, wxConvLocal).c_str());
    wxLogDebug(wxT("GetDefaultPlayDevice() default output %d"), apiinfo->defaultOutputDevice);
    const PaDeviceInfo* devinfo = Pa_GetDeviceInfo(apiinfo->defaultOutputDevice);
@@ -308,8 +313,11 @@
 
 wxString DevicePrefs::GetDefaultRecordDevice(int index)
 {
-   const struct PaHostApiInfo *apiinfo;   /* info on this API */
-   apiinfo = Pa_GetHostApiInfo(index);   // get info on API
+   if (index < 0 || index >= Pa_GetHostApiCount()) {
+      return wxEmptyString;
+   }
+
+   const struct PaHostApiInfo *apiinfo = Pa_GetHostApiInfo(index);   // get info on API
    wxLogDebug(wxT("GetDefaultRecordDevice(): HostAPI index %d, name %s"), index, wxString(apiinfo->name, wxConvLocal).c_str());
    wxLogDebug(wxT("GetDefaultRecordDevice() default input %d"), apiinfo->defaultInputDevice);
    const PaDeviceInfo* devinfo = Pa_GetDeviceInfo(apiinfo->defaultInputDevice);
